src/metabase/api/table.clj

```diff
@@ -516,10 +516,14 @@
       {:status 200
        :body   (:id model)})
     (catch Throwable e
-      {:status (or (-> e ex-data :status-code)
-                   500)
-       :body   {:message (or (ex-message e)
-                             (tru "There was an error uploading the file"))}})
+      (if (::upload/schema-mismatch? (ex-data e))
+        {:status (-> e ex-data :status-code)
+         :body   (merge {:message (ex-message e)}
+                        (select-keys (ex-data e) [:extra-columns :missing-columns]))}
+        {:status (or (-> e ex-data :status-code)
```

Comment:
Since `ex-data` is called twice and `ex-message` is called twice, I wonder if it isn't worth pulling them both out in a `let`.

Reply:
Yep. This isn't actually ready for review yet, I created this PR because I wanted to park it with a public PR page while I waited for input from Ryan on the FE. I generally add the team as reviewers when it's ready for review